Task queue

kabachook opened the pull request:
- Redis' list as queue (`lpush` and `brpop`)
- Test coverage ~=65%

kabachook committed:
Task queue

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+/**
+ * Json message parser for Queue
+ */
+class JsonParser extends Parser {
```
лучше вынести в отдельные файлы. И в начале файла нужно подробно описать, что это, зачем он нужен и как работает

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+    redisClient = null
+  }) {
+    super(queueName);
+    this.prefix = prefix;
```
отделяй смысловые части пустыми строками

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+  /**
+   * Create a RedisQueue.
+   * @param {string} queueName Queue name.
+   * @param {string} [prefix='queue:'] Prefix.
```
префикс чего? для чего нужен

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+  }) {
+    super(queueName);
+    this.prefix = prefix;
+    this._fullName = prefix + queueName;
```
свойства класса надо описать в комментах

neSpecc commented on task-queue/lib/queue.test.js:
```diff
@@ -0,0 +1,89 @@
+  };
+  let queue;
+
+  beforeAll(() => {
```
все методы и смысловые части надо задокументировать

neSpecc commented on task-queue/lib/queue.test.js:
```diff
@@ -0,0 +1,89 @@
+
+const REDIS_URL = 'redis://127.0.0.1:6379';
+const QUEUE_NAME = 'test';
+const QUEUE_TIMEOUT = 30;
```
все константы надо задокументировать

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+ */
+class JsonParser extends Parser {
+  /**
+   * Prepare message
```
надо пояснить, какие действия входят в prepare

khaydarov commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+   * @throws {Error} Not implemented.
+   * @returns {void}
+   */
+  parse() {
```
функция не принимает ничего, а в JSDoc ты указал msg

khaydarov commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+   * Prepare message
+   * @param {object} msg Message.
+   * @returns {string} Prepared message.
+   * @throws {Error}
```
@throws {ParserError}

khaydarov commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+  static parse(msg) {
+    try {
+      return JSON.parse(msg);
+    } catch (e) {
```
e - нигде не используется
и добавь throws

khaydarov commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+}
+
+/**
+ * Queue base class.
```
Опиши подробнее класс. Зачем он нужен, как использовать и какие задачи решает

khaydarov commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,199 @@
+ */
+class Queue extends EventEmitter {
+  /**
+   *
```
лишнее

kabachook committed:
Code refactor

- Remove EventEmitter inheritance
- Split Queue and Parser to two files
- More comments
- Remove timeout argument from RedisQueue

neSpecc commented on task-queue/lib/parser.js:
```diff
@@ -0,0 +1,80 @@
+/**
+ * This file is corresponding to message parsers.
+ *
+ * Whenever queue sends ot receives a message, it gets processed using Parsers's methods.
```
```suggestion
 * Whenever Queue sends or receives a message, it gets processed using Parsers's methods.
```

neSpecc replied:
и что значит queue sends a message? очередь разве может что-то отправлять куда-то? туда можно положить задачу и можно взять

kabachook replied:
Имелось в виду класс. Не написал большую букву

neSpecc commented on task-queue/lib/parser.js:
```diff
@@ -0,0 +1,80 @@
+ * This file is corresponding to message parsers.
+ *
+ * Whenever queue sends ot receives a message, it gets processed using Parsers's methods.
+ * For example, if we reiceve a message, we call `parser.parse(msg)`.
```
```suggestion
 * For example, if we receive a message, we call `parser.parse(msg)`.
```

neSpecc commented on task-queue/lib/parser.js:
```diff
@@ -0,0 +1,80 @@
+ * Whenever queue sends ot receives a message, it gets processed using Parsers's methods.
+ * For example, if we reiceve a message, we call `parser.parse(msg)`.
+ * This allows us to use many data formats like json, msgpack, binary, etc.
+ * You can write your own parser just by inheriting from `Parser` and implementing `prepare` and `parse`.
```
сразу надо описать, что делают prepare и parse

neSpecc commented on task-queue/lib/parser.js:
```diff
@@ -0,0 +1,80 @@
+class Parser {
+  /**
+   * Prepare message.
+   * Here you convert a message TO specified data format.
```
```suggestion
   * Here you convert a message to specified data format.
```

neSpecc commented on task-queue/lib/parser.js:
```diff
@@ -0,0 +1,80 @@
+
+  /**
+   * Parse message.
+   * Here you convert a message FROM specified data format.
```
```suggestion
   * Here you convert a message from specified data format.
```

neSpecc commented on task-queue/lib/parser.test.js:
```diff
@@ -0,0 +1,27 @@
+  ParserError
+} = require('./parser');
+
+describe('JsonParser', () => {
```
опиши кратко, что проверяет этот тест

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,154 @@
+/**
+ * This file is corresponding to queues.
```
что это значит?

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,154 @@
+ * @property {Parser} parser Parser used by queue to convert messages to specified data format.
+ * @property {IORedis.Redis} dbClient Redis client.
+ */
+class RedisQueue extends Queue {
```
лучше вынести в отдельный файл

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,154 @@
+  }) {
+    super(queueName);
+
+    this.queueName = queueName;
```
не задокументированы свойства

kabachook replied:
В описании класса написал

neSpecc commented on task-queue/lib/queue.js:
```diff
@@ -0,0 +1,154 @@
+
+  /**
+   * Sends message to queue.
+   * @param {object | Array} msg Message to send.
```
если известна структура объекта, надо ее указать. И у массива указать тип элементов

kabachook replied:
Тут msg - это любой обьект, который можно преобразовать в json строчку

neSpecc committed:
Fix typo

Co-Authored-By: kabachook <beltyukovdanil@gmail.com>

neSpecc committed:
Fix typo

Co-Authored-By: kabachook <beltyukovdanil@gmail.com>

neSpecc committed:
Fix typo

Co-Authored-By: kabachook <beltyukovdanil@gmail.com>

neSpecc committed:
Fix typo

Co-Authored-By: kabachook <beltyukovdanil@gmail.com>

kabachook committed:
Fix jsdoc comments

kabachook committed:
Separate file for JsonParser and test description

kabachook committed:
Separete file for RedisQueue and more comments

kabachook committed:
More comments and rename Queue methods

neSpecc commented on task-queue/lib/parser.js:
```diff
@@ -0,0 +1,54 @@
+ * for example, a JS object to supported data format of database.
+ * For instanse Redis does not support JSON format, so we convert any JSONable JS data type to string.
+ *
+ * `prepare` converts a message to specified format, `parse` converts from specified format to JS object.
```
```suggestion
 * `prepare` — converts a message to specified format, 
 * `parse` — converts from specified format to JS object.
```

neSpecc committed:
Fix typo

Co-Authored-By: kabachook <beltyukovdanil@gmail.com>
